### PR TITLE
1889: MRU picker can contain characters not in enabled character suite

### PR DESCRIPTION
**Pull request by chrismiceli**
filter out characters not present in character picker

### CODE REVIEW

**Comment by chrismiceli** [comment]
Sandbox here https://www.pgdp.org/~chrismiceli/c.branch/issue/1889/

**Review by srjfoo — APPROVED** [review]
Looks good to me.

I tested by adding Polytonic Greek to a project, proofing a page and adding Greek using the character picker, verifying that the MRU had the Greek characters, saving, removing the Polytonic Greek charsuite and verifying that only the Greek characters were removed from the MRU.

If there's any other way it needs to be tested, let me know.

**Comment by chrismiceli** [comment]
> Looks good to me.
> 
> I tested by adding Polytonic Greek to a project, proofing a page and adding Greek using the character picker, verifying that the MRU had the Greek characters, saving, removing the Polytonic Greek charsuite and verifying that only the Greek characters were removed from the MRU.
> 
> If there's any other way it needs to be tested, let me know.

You could use browser developer tools to modify localStorage but what you did is the real world scenario. Thanks for testing!

File: tools/proofers/character_selector.js [hunk 1]
```diff
@@ -38,6 +38,16 @@ $(function () {
...
+    var pickerChars = {};
+    $('.picker').each(function pickerEach() {
+        pickerChars[$(this).text()] = true;
+    });
+
+    // filter out invalid characters
+    mru = mru.filter(function mruFilter(mruCharacter) {
+        return pickerChars[mruCharacter.character] === true;
+    });
+
```
Comment (70ray) [inline tools/proofers/character_selector.js:50]: The functions could be anonymous.

**Review by 70ray — APPROVED** [review]

File: tools/proofers/character_selector.js (hunk 1 above)
Comment (chrismiceli) [reply tools/proofers/character_selector.js:50]: Sure, I am used to naming them for better stack traces, but this code isn't too complex.

**Comment by srjfoo** [comment]
Re-tested after the code update, with different pickersets but the same results: MRU loses the characters that are in a character suite once it's removed. Re-did test with custom characters and the same results. Once the Project custom characters are removed, any instances of them in the MRU are also removed.

**Pull request merged**